TemperatureMeter.c

```diff
@@ -0,0 +1,154 @@
+         continue;
+
+      int m = 0;
+      for (const sensors_feature *feature = sensors_get_features(chip, &m); feature; feature = sensors_get_features(chip, &m)) {
+         if (feature->type != SENSORS_FEATURE_TEMP)
+            continue;
+
+         if (feature->number != core)
+            continue;
+
+         const sensors_subfeature *sub_feature = sensors_get_subfeature(chip, feature, type);
+         if (sub_feature) {
+            double local_temp;
+            if (sensors_get_value(chip, sub_feature->number, &local_temp) == 0) {
+               *temp = this->pl->settings->fahrenheitTemperature ? local_temp * 9 / 5 + 32 : local_temp;
+               return 0;
+            }
+         }
+      }
+   }
+
+   return -ENOENT;
+}
+
+int TemperatureMeter_attributes[] = {
+   TEMPERATURE_COOL,
+   TEMPERATURE_MEDIUM,
+   TEMPERATURE_HOT,
+};
+
+static void TemperatureMeter_updateValues(Meter* this, char* buffer, int len) {
+   double temp;
+
+   // Show only package temperature in bars meter mode
+   if (get_cpu_temp(this, &temp, 0, SENSORS_SUBFEATURE_TEMP_INPUT) < 0) {
+      xSnprintf(buffer, len, "N/A");
+      return;
+   }
+   get_cpu_temp(this, &this->total, 0, SENSORS_SUBFEATURE_TEMP_MAX);
+
+   this->values[0] = temp;
+   if (this->pl->settings->fahrenheitTemperature)
+      xSnprintf(buffer, len, "%d°F", (int)temp);
+   else
+      xSnprintf(buffer, len, "%d°C", (int)temp);
+}
+
+static void TemperatureMeter_display(Object* cast, RichString* out) {
+   Meter* this = (Meter*)cast;
+   int sensors_amount = get_cpu_sensors_amount();
+   double temp;
+
+   if (get_cpu_temp(this, &temp, 0, SENSORS_SUBFEATURE_TEMP_INPUT) < 0 || sensors_amount == 0) {
+      RichString_append(out, CRT_colors[METER_VALUE], "n/a");
+      return;
+   }
+
+   int core_count_begin = sensors_amount == 1 ? 0 : 1;
+
+   for (int core = core_count_begin; core < sensors_amount; core++) {
+      get_cpu_temp(this, &temp, core, SENSORS_SUBFEATURE_TEMP_INPUT);
+
+      int temp_medium = this->pl->settings->fahrenheitTemperature ? 140 : 60;
+      int temp_hot    = this->pl->settings->fahrenheitTemperature ? 158 : 70;
+
+      // convert double to integer
+      int temperature = (int)temp;
+
+      // choose the color for the temperature
+      int tempColor;
+      if      (temperature < temp_medium) tempColor = CRT_colors[TEMPERATURE_COOL];
+      else if (temperature < temp_hot)    tempColor = CRT_colors[TEMPERATURE_MEDIUM];
+      else                                tempColor = CRT_colors[TEMPERATURE_HOT];
+
+      // output the temperature
+      char buffer[4];
+      snprintf(buffer, 4, "%3d", temperature);
+      RichString_append(out, tempColor, buffer);
+      if (this->pl->settings->fahrenheitTemperature)
+         RichString_append(out, CRT_colors[METER_TEXT], "°F ");
+      else
+         RichString_append(out, CRT_colors[METER_TEXT], "°C ");
+   }
+}
```

Comment:
Remodel the `get_cpu_sensors_amount()` function to both count the CPUs AND retrieve status information for each core at once. The returned structure should best be indexed in a wait for the actual information display step to directly lookup the proper information without additional searching in the returned data, but direct array access. Current O(n³) complexity is deadly for systems with many CPU cores when basically O(n) or O(n*log(n)) should easily be achievable.

Reply:
Reworked to structure with dynamic array.